MIGRATION.md

```diff
@@ -0,0 +1,130 @@
+# Migrating to InstantSearch v4
+
+This latest version of InstantSearch is focused mostly around Federated Search, as well as bundle size. To achieve this, we have made some changes which impact your apps. These changes are detailed here:
+
+## Federated Search (multi-index)
+
+If you were already using federated search, likely via synchronizing two InstantSearch indices via `searchFunction`, you can now rejoice that this is simpler. There's now a new `index` widget, upon which you can attach more widgets. An example use case is the following:
+
+```js
+const search = instantsearch({ indexName: 'primary', /* ... */ });
+
+search.addWidgets([
+  searchBox(),
+  hits(),
+  index({ indexName: 'secondary' }).addWidgets([
+    searchBox(),
+    hits(),
+  ])
+]);
+```
+
+A more detailed guide on this topic will follow once this version is out of beta stage.
+
+## Routing
+
+Even if you are not using multi-index, the UiState has changed shape; what used to be:
+
+```json
+{
+  "query": "value",
+  "page": 5
+}
+```
+
+Is now:
+
+```json
+{
+  "indexName": {
+    "query": "value",
+    "page": 5
+  }
+}
+```
+
+If you are using the default stateMapping with the current version, you can replace that with `singleIndexStateMapping('yourIndexName')` (import from `instantsearch.js/es/lib/stateMappings`).
+
+If you were using a custom stateMapping, you need to loop over the outer level of index, and add this extra level back in `routeToState`. You can check the [source](https://github.com/algolia/instantsearch.js/blob/next/src/lib/stateMappings/singleIndex.ts) for reference on how to implement this.
```

Comment:
What do you think of having a small example for before/after?

```js
// before
const stateMapping = {
  stateToRoute(uiState) {
    return {
      query: uiState.query,
      page: uiState.page,
      // ...
    };
  },

  routeToState(routeState) {
    return {
      query: routeState.query,
      page: routeState.page,
      // ...
    };
  },
};

// after
const stateMapping = {
  stateToRoute(uiState) {
    const indexUiState = uiState[indexName];
    return {
      query: indexUiState.query,
      page: indexUiState.page,
      // ...
    };
  },

  routeToState(routeState) {
    return {
      [indexName]: {
        query: routeState.query,
        page: routeState.page,
        // ...
      },
    };
  },
};
```

Reply:
makes sense, I'll add this example